nft_ingester/src/backfiller.rs

```diff
@@ -0,0 +1,702 @@
+        // Get the maximum sequence number that has been backfilled, and use
+        // that for the starting sequence number for backfilling.
+        let mut query = backfill_items::Entity::find()
+            .select_only()
+            .column(backfill_items::Column::Seq)
+            .filter(backfill_items::Column::Backfilled.eq(true))
+            .filter(backfill_items::Column::Tree.eq(tree))
+            .group_by(backfill_items::Column::Tree)
+            .build(DbBackend::Postgres);
+
+        query.sql = query.sql.replace(
+            "SELECT \"backfill_items\".\"seq\"",
+            "SELECT MAX (\"backfill_items\".\"seq\")",
+        );
+
+        // Debug.
+        //println!("{}", query.to_string());
+
+        let start_seq_vec = MaxSeqItem::find_by_statement(query).all(&self.db).await?;
+        let start_seq = if start_seq_vec.len() > 0 {
+            start_seq_vec[0].max
+        } else {
+            0
+        };
+
+        // Debug.
+        //println!("MAX: {}", start_seq);
+
+        // Get all rows for the tree that have not yet been backfilled.
+        let mut query = backfill_items::Entity::find()
+            .select_only()
+            .column(backfill_items::Column::Seq)
+            .column(backfill_items::Column::Slot)
+            .filter(backfill_items::Column::Seq.gte(start_seq))
+            .filter(backfill_items::Column::Tree.eq(tree))
+            .order_by_asc(backfill_items::Column::Seq)
+            .build(DbBackend::Postgres);
+
+        query.sql = query.sql.replace("SELECT", "SELECT DISTINCT");
+        let rows = SimpleBackfillItem::find_by_statement(query)
+            .all(&self.db)
+            .await?;
+        let mut gaps = vec![];
+
+        // Look at each pair of subsequent rows, looking for a gap in sequence number.
+        for (prev, curr) in rows.iter().zip(rows.iter().skip(1)) {
+            if curr.seq == prev.seq {
+                let message = format!(
+                    "Error in DB, identical sequence numbers with different slots: {}, {}",
+                    prev.slot, curr.slot
+                );
+                println!("{}", message);
+                return Err(DbErr::Custom(message));
+            } else if curr.seq - prev.seq > 1 {
+                gaps.push(GapInfo::new(prev.clone(), curr.clone()));
+            }
+        }
+
+        // Get the max sequence number if any rows were returned from the query.
+        let opt_max_seq = rows.last().map(|row| row.seq);
+
+        Ok((opt_max_seq, gaps))
+    }
+
+    // Similar to `plugGaps()` in `backfiller.ts`.
+    async fn plug_gap(&mut self, gap: &GapInfo, tree: &[u8]) -> Result<(), IngesterError> {
+        // TODO: This needs to make sure all slots are available otherwise it will partially
+        // fail and redo the whole backfill process.  So for now checking the max block before
+        // looping as a quick workaround.
+        let _ = self
```

Comment:
Yeah I need a cleaner way.  The reason this TODO is here is because here's an example:

I need to get blocks 33, 34, and 35 for backfilling so I have code right below this that backfills for smallest block to largest block.  If 33 is finalized but 34 isn't, I'll end up successfully backfilling block 33 but then failing on block 34 until it is finalized.  So I did a check to make sure block 35 is finalized before moving on.

What I want to do is walk through the blocks in reverse order (in this example 35 to 33), and that might actually result in less database writes if we only update when `seq` > existing `seq`.  So it might allow us to favor newer data.  But I haven't thought all that through yet, and again wanted to wait until we decided if we are switching to transaction based gap detection which might result in only getting one block at a time to process.